### ISSUE TITLE
SPeDE assigns spectra to multiple cluster/group and crashes with an InvalidIndexError

### ISSUE THREAD

**Issue by cpauvert** [LM-UGent/SPeDE#14]
Hi,
I am running into an error with six spectra: 
* Bifidobacterium_animalis
* 2 Staphylococcus_hominis
* Bacteroides_eggerthii
* 2 Bacteroides_acidifaciens

The preprocessing was fine, however, I ran into the following error below:
```
python3 /SPeDE/Spectrum_Processing/SPeDE.py /SPeDE/GUI/ppmc_interval_index.csv test_mini/ .
Sys: Reading all spectra...
Sys: Reading and preprocessing PKL files...
Sys: Reading FMS files...
Sys: Reading interval list...
Sys: Creating SpectrumMatcher...
Sys: Calculating uniqueness matrix...
Sys: Start calculating using 8 processess...
Sys: Preparing clustering data...
Sys: Generating cluster reference list...
Sys: Preparing dice data...
Sys: Generating dice matrix...
Sys: Generating dice referlist...
Sys: Writing clustering...
                                     BRUKER_SCORE  BRUKER_TOPHIT  STRAIN
SOURCE_FILE                                                             
Bifidobacterium_animalis_1_MiBC_C11           NaN            NaN     NaN
Staphylococcus_hominis_1_PIBAC_E5             NaN            NaN     NaN
Staphylococcus_hominis_2_PIBAC_E6             NaN            NaN     NaN
Bacteroides_eggerthii_1_A5                    NaN            NaN     NaN
Bacteroides_acidifaciens_2_MiBC_A12           NaN            NaN     NaN
Bacteroides_acidifaciens_1_MiBC_A11           NaN            NaN     NaN
Staphylococcus_hominis_1_PIBAC_E5  <-         NaN            NaN     NaN
Staphylococcus_hominis_2_PIBAC_E6  <-         NaN            NaN     NaN
Traceback (most recent call last):
  File "/SPeDE/Spectrum_Processing/SPeDE.py", line 123, in <module>
    main(args.intervals, args.project_directory, args.output_directory, args.peaks, args.density, args.local,args.cutoff, args.validate,
  File "/SPeDE/Spectrum_Processing/SPeDE.py", line 70, in main
    cluster.write_clustering(output_format)
  File "/SPeDE/Spectrum_Processing/../Spectrum_Processing/cluster.py", line 195, in write_clustering
    format_CSV(self.refer_array, self.group_array, os.path.join(self.output_directory ,self.out_name), self.provider)
  File "/SPeDE/Spectrum_Processing/../Spectrum_Processing/postProcess.py", line 106, in format_CSV
    result = pd.concat([data, qualities, table], axis=1,join="inner",sort=False)
  File "/usr/local/lib/python3.10/dist-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pandas/core/reshape/concat.py", line 381, in concat
    return op.get_result()
  File "/usr/local/lib/python3.10/dist-packages/pandas/core/reshape/concat.py", line 612, in get_result
    indexers[ax] = obj_labels.get_indexer(new_labels)
  File "/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py", line 3905, in get_indexer
    raise InvalidIndexError(self._requires_unique_msg)
pandas.errors.InvalidIndexError: Reindexing only valid with uniquely valued Index objects
```

Have you ever ran into a similar issues?

Pandas is complaining that the merge of the DataFrame cannot happen because some index values appears multiple times (for _S. hominis_ it seems , with arrows). 

Two additional information:
1. when just the two _S. hominis_ are ran together, SPeDE complains that there is no valid spectra.
2. when just the two _S. hominis_ are ran with a _B. acidifaciens_, it runs fine but the two _S. hominis_ are flagged as ORANGE quality.

Could it be that the ORANGE quality spectra generate trouble?

Best,

**Comment by cpauvert** [comment]
A follow-up on this issue.
**TL;DR**: SPeDE does not seem to know where to place spectrum that could belong to two groups in this case _S. hominis_ to _B. animalis_ and to _B. acidifaciens_ (which is a strange thing to do anyway), have you had experiences with this behavior already @edepoort @esdpoort ?

It appears that this is not an issue with the txt file format (using `-o txt`), S. hominis E5 and E6 are NOT put twice (see the output below):
```
GROUP Bifidobacterium_animalis_1_MiBC_C11.tab 0
GROUP Bacteroides_eggerthii_1_A5.tab 1
GROUP Bacteroides_acidifaciens_2_MiBC_A12.tab 2
REFER Bacteroides_acidifaciens_1_MiBC_A11.tab Bacteroides_acidifaciens_2_MiBC_A12.tab
REFER Staphylococcus_hominis_1_PIBAC_E5.tab Bifidobacterium_animalis_1_MiBC_C11.tab
REFER Staphylococcus_hominis_1_PIBAC_E5.tab Bacteroides_acidifaciens_2_MiBC_A12.tab
REFER Staphylococcus_hominis_2_PIBAC_E6.tab Bifidobacterium_animalis_1_MiBC_C11.tab
REFER Staphylococcus_hominis_2_PIBAC_E6.tab Bacteroides_acidifaciens_2_MiBC_A12.tab
``` 

When printing the `self.refer_array` in the `write_clustering` function in `cluster.py`, all the comparisons outlined are fine (see below), meaning the issue is in the printing/formatting but ONLY in csv format.

```
# Here the order is spectrum, reference_spectrum, meaning A11 should belong to A12 group, E5 should belong to two different group C11 and A12
[
 ['Bacteroides_acidifaciens_1_MiBC_A11.tab', 'Bacteroides_acidifaciens_2_MiBC_A12.tab'],
 ['Staphylococcus_hominis_1_PIBAC_E5.tab', 'Bifidobacterium_animalis_1_MiBC_C11.tab'],
 ['Staphylococcus_hominis_1_PIBAC_E5.tab', 'Bacteroides_acidifaciens_2_MiBC_A12.tab'],
 ['Staphylococcus_hominis_2_PIBAC_E6.tab', 'Bifidobacterium_animalis_1_MiBC_C11.tab'],
 ['Staphylococcus_hominis_2_PIBAC_E6.tab', 'Bacteroides_acidifaciens_2_MiBC_A12.tab']
]
```

When printing `data` in `postProcess.py` before the faulty line 106 (mentioned in the log in the first post) that performs the concatenation of the Pandas DataFrame `data`, `qualities` and `table`, there is already two additional lines with S. hominis: one that belongs to reference group number 2 one that belongs to group number 0 :

```
                                                             SOURCE_FILE REFERENCE  REFERENCE_NUMBER  GROUP
SOURCE_FILE                                                                                                
Bifidobacterium_animalis_1_MiBC_C11  Bifidobacterium_animalis_1_MiBC_C11       Yes                 0      0
Staphylococcus_hominis_1_PIBAC_E5      Staphylococcus_hominis_1_PIBAC_E5        No                 0      0
Staphylococcus_hominis_2_PIBAC_E6      Staphylococcus_hominis_2_PIBAC_E6        No                 0      0
Bacteroides_eggerthii_1_A5                    Bacteroides_eggerthii_1_A5       Yes                 1      1
Bacteroides_acidifaciens_2_MiBC_A12  Bacteroides_acidifaciens_2_MiBC_A12       Yes                 2      2
Bacteroides_acidifaciens_1_MiBC_A11  Bacteroides_acidifaciens_1_MiBC_A11        No                 2      2
Staphylococcus_hominis_1_PIBAC_E5      Staphylococcus_hominis_1_PIBAC_E5        No                 2      2
Staphylococcus_hominis_2_PIBAC_E6      Staphylococcus_hominis_2_PIBAC_E6        No                 2      2
```

**Comment by esdpoort** [comment]
Thank you  @cpauvert for documenting this issue. The issue was indeed caused by duplicate spectra present in the `refer_array`, namely the two _S hominis_ spectra which were each matched to two references. This duplication causes pandas to refuse the merge of the dataframes, as you observed. I implemented a fix for this issue by dropping duplicates before merging the different DataFrames and keeping just the first occurrence of the spectrum. 

This issue is something we have not seen before and I assume it is simply because we only very rarely encounter spectra marked as 'orange' or 'average quality' in our datasets. In our experience, spectra are either 'green' (good quality) or 'red' (bad quality). If you are using a different strategy for peak calling than the R script provided with SPeDE, I think it would be worthwhile to look into the peak calling process and check if the spectra that SPeDE marks as 'orange' or 'average quality' are truly that. In addition, you could try modifying the S/N thresholds for determining the quality of a spectrum (currently hard-coded in  `provider.py`).

**Comment by cpauvert** [comment]
Thanks @esdpoort for the fix! I re ran the dataset and I don't have the issue, so thanks!
I was using the same procedure as recommended by the preprocessing in SPeDE.
I'll explore the consequences of this fix later next year.
Best,
